// redux-saga-sc/lib/actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.socketRequest = exports.socketEmit = exports.REQUEST = exports.EMIT = undefined;

var _utils = require('./utils');

var EMIT = exports.EMIT = (0, _utils.sym)('EMIT');
var REQUEST = exports.REQUEST = (0, _utils.sym)('REQUEST');

var socketEmit = exports.socketEmit = function socketEmit(payload) {
  var event = arguments.length <= 1 || arguments[1] === undefined ? 'dispatch' : arguments[1];
  var autoReconnectOptions = arguments[2];
  return {
    type: EMIT,
    event: event,
    autoReconnectOptions: autoReconnectOptions,
    payload: payload
  };
};

var socketRequest = exports.socketRequest = function socketRequest(payload) {
  var event = arguments.length <= 1 || arguments[1] === undefined ? 'dispatch' : arguments[1];
  var autoReconnectOptions = arguments[2];
  var timeout = arguments[3];
  return {
    type: REQUEST,
    event: event,
    autoReconnectOptions: autoReconnectOptions,
    timeout: timeout,
    payload: payload
  };
};
